operators/src/opencl/cl_program.rs

```diff
@@ -0,0 +1,578 @@
+        v.resize_with(output_raster_types.len(), || None);
+
+        Self {
+            input_rasters: vec![None; input_raster_types.len()],
+            output_rasters: v,
+            input_raster_types,
+            output_raster_types,
+        }
+    }
+
+    pub fn set_input_raster(&mut self, idx: usize, raster: &'a TypedRaster2D) -> Result<()> {
+        ensure!(
+            idx < self.input_raster_types.len(),
+            error::CLProgramInvalidRasterIndex
+        );
+        ensure!(
+            raster.raster_data_type() == self.input_raster_types[idx],
+            error::CLProgramInvalidRasterDataType
+        );
+        self.input_rasters[idx] = Some(raster);
+        Ok(())
+    }
+
+    pub fn set_output_raster(&mut self, idx: usize, raster: &'a mut TypedRaster2D) -> Result<()> {
+        ensure!(
+            idx < self.input_raster_types.len(),
+            error::CLProgramInvalidRasterIndex
+        );
+        ensure!(
+            raster.raster_data_type() == self.output_raster_types[idx],
+            error::CLProgramInvalidRasterDataType
+        );
+        self.output_rasters[idx] = Some(raster);
+        Ok(())
+    }
+}
+
+/// Allows running kernels on different inputs and outputs
+pub struct CompiledCLProgram {
+    kernel: Kernel,
+    iteration_type: IterationType,
+    input_raster_types: Vec<RasterDataType>,
+    output_raster_types: Vec<RasterDataType>,
+    output_buffers: Vec<OutputBuffer>,
+}
+
+impl CompiledCLProgram {
+    pub fn new(
+        kernel: Kernel,
+        iteration_type: IterationType,
+        input_raster_types: Vec<RasterDataType>,
+        output_raster_types: Vec<RasterDataType>,
+    ) -> Self {
+        Self {
+            kernel,
+            iteration_type,
+            input_raster_types,
+            output_raster_types,
+            output_buffers: vec![],
+        }
+    }
+
+    pub fn params<'a>(&self) -> CLProgramParameters<'a> {
+        CLProgramParameters::new(
+            self.input_raster_types.clone(),
+            self.output_raster_types.clone(),
+        )
+    }
+
+    fn add_output_buffer<T>(
+        &mut self,
+        idx: usize,
+        raster: &Raster2D<T>,
+        f: fn(Buffer<T>) -> OutputBuffer,
+    ) -> Result<()>
+    where
+        T: Pixel + ocl::OclPrm,
+    {
+        let buffer = Buffer::<T>::builder()
+            .queue(self.kernel.default_queue().expect("checked").clone())
+            .len(raster.data_container.len())
+            .build()
+            .context(error::OCL)?;
+
+        self.kernel
+            .set_arg(format!("OUT{}", idx), &buffer)
+            .context(error::OCL)?;
+
+        self.output_buffers.push(f(buffer));
+
+        Ok(())
+    }
+
+    fn set_arguments(&mut self, params: &CLProgramParameters) -> Result<()> {
+        ensure!(
+            params.input_rasters.iter().all(Option::is_some),
+            error::CLProgramUnspecifiedRaster
+        );
+        self.output_buffers.clear();
```

Comment:
Why are only the output buffers cleared?

Reply:
because the input buffers are filled with data from the host slices